Sources/SwiftFusion/Image/Patch.swift

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import TensorFlow
+
+extension Tensor where Scalar == Double {
+  /// Returns the patch of `self` at `region`, interpolating pixels using `interpolation`.
+  ///
+  /// Coordinate system used in `region`: The pixel at `self[j, i, ...]` is the square bounded by
+  /// the corners with coordinates `(i, j)` and `(i + 1, j + 1)`. The center of the pixel is at
+  /// coorindate `(i + 0.5, j + 0.5)`.
+  ///
+  /// For example, `(0, 0)` is the top left of the image and `(width, height)` is the bottom right
+  /// of the image.
+  ///
+  /// - Parameters:
+  ///   - self: an image, with shape `[height, width, channelCount]`.
+  ///   - region: the center, orientation, and size of the patch to extract. The center is
+  ///     specified in the coordinate system described above, and the size is specified in units
+  ///     of pixels.
+  @differentiable
+  public func patch(
+    at region: OrientedBoundingBox,
+    interpolation: @differentiable (Tensor<Scalar>, Vector2) -> Tensor<Scalar>
+  ) -> Tensor<Scalar> {
+    precondition(self.shape.count == 3, "image must have shape height x width x channelCount")
+    let patchShape: TensorShape =
+      withoutDerivative(at: [Int(region.size.y), Int(region.size.x), self.shape[2]])
```

Comment:
They are floating point now because I store them as `Vector2` in the `OrientedBoundingBox` `size` field. But I do think it makes more sense to make them integers, so I'll update `OrientedBoundingBox` to have integer size fields instead of floating point size fields.